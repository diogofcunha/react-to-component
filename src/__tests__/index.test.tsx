import deepEquals from "fast-deep-equal";
import React from "react";
import { cleanup, render, waitForElement } from "react-testing-library";

import ToComponent, { Props } from "..";

type TestProps = Props<any, any>;

// tslint:disable-next-line:no-console
console.error = x => {
  throw new Error(x);
};

describe("<ToComponent />", () => {
  afterEach(cleanup);

  const setup = (
    p: Pick<TestProps, "original" | "params">,
    parametersDidChange?: (prev: any[], cur: any[]) => boolean
  ) => {
    const props: TestProps = {
      children: jest.fn().mockImplementation(({ error, loading }) => {
        if (loading) {
          return "Loading";
        }

        if (error) {
          return error.message;
        }

        return "Loaded";
      }),
      parametersDidChange,
      ...(p as any)
    };

    return {
      props,
      ...render(<ToComponent {...props} />)
    };
  };

  const waitFor = (ms: number) =>
    new Promise(r => {
      setTimeout(() => {
        r();
      }, ms);
    });

  describe("Sync apis", () => {
    test("should call children with the correct props on success", async () => {
      const original = (name: string, age: number) =>
        `I am ${name} and I am ${age} years old`;
      const params = ["Diogo", 27];

      const { props, getByText } = setup({ original, params });

      await waitForElement(() => getByText("Loaded"));

      expect(props.children).toMatchSnapshot();
    });

    test("should call children with the correct props on failure", async () => {
      const error = new Error("Trolololololol");

      const original = () => {
        throw error;
      };

      const params = ["Diogo", 27];

      const { props, getByText } = setup({ original, params });

      await waitForElement(() => getByText(error.message));

      expect(props.children).toMatchSnapshot();
    });

    test("should use custom parametersDidChange when provided", async () => {
      const original = (
        name: string,
        age: number,
        log: (res: string) => void
      ) => {
        const res = `I am ${name} and I am ${age} years old`;
        log(res);

        return res;
      };

      const params = ["Diogo", 27, () => {}];

      const parametersDidChange = (prevParams: any[], nextParams: any[]) => {
        return !deepEquals(prevParams.slice(0, 2), nextParams.slice(0, 2));
      };

      const { props, rerender, getByText } = setup(
        { original, params },
        parametersDidChange
      );

      await waitForElement(() => getByText("Loaded"));

      rerender(<ToComponent {...props} params={["Diogo", 27, () => {}]} />);

      await waitForElement(() => getByText("Loaded"));

      expect(props.children).toMatchSnapshot();
    });
  });

  describe("Async apis", () => {
    test("should call children with the correct props on success", async () => {
      const original = async (name: string, age: number) => {
        await waitFor(50);

        return `I am ${name} and I am ${age} years old`;
      };

      const params = ["Diogo", 27];

      const { props } = setup({ original, params });

      await waitFor(75);

      expect(props.children).toMatchSnapshot();
    });

    test("should call children with the correct props on failure", async () => {
      const error = new Error("trolololol");

      const original = async () => {
        await waitFor(50);

        throw error;
      };

      const { props, getByText } = setup({ original, params: ["Diogo", 27] });

      await waitForElement(() => getByText(error.message));

      expect(props.children).toMatchSnapshot();
    });

    describe("async flow problems", () => {
      test("should call children with the correct props on success", async () => {
        const original = async (name: string, age: number) => {
          if (name === "Fonseca") {
            await waitFor(100);
          } else {
            await waitFor(50);
          }

          return `I am ${name} and I am ${age} years old`;
        };

        const { props, rerender, getByText } = setup({
          original,
          params: ["Fonseca", 27]
        });

        rerender(<ToComponent {...props} params={["Diogo", 27]} />);

        await waitFor(125);
        await waitForElement(() => getByText("Loaded"));

        expect(props.children).toMatchSnapshot();
      });

      test("should call children with the correct props on success and ignore old errors", async () => {
        const original = async (name: string, age: number) => {
          if (name === "Diogo") {
            return `I am ${name} and I am ${age} years old`;
          } else {
            await waitFor(50);
            throw Error("trololo");
          }
        };

        const { props, rerender, getByText } = setup({
          original,
          params: ["Fonseca", 27]
        });

        rerender(<ToComponent {...props} params={["Diogo", 27]} />);

        await waitFor(75);
        await waitForElement(() => getByText("Loaded"));

        expect(props.children).toMatchSnapshot();
      });

      test("should call children with the correct props on failure", async () => {
        const error = new Error("trolololololololol");

        const original = async (name: string, age: number) => {
          if (name === "Fonseca") {
            await waitFor(100);

            return `I am ${name} and I am ${age} years old`;
          } else {
            await waitFor(50);

            throw new Error("trolololololololol");
          }
        };

        const { props, rerender, getByText } = setup({
          original,
          params: ["Fonseca", 27]
        });

        rerender(<ToComponent {...props} params={["Diogo", 27]} />);

        await waitFor(125);
        await waitForElement(() => getByText(error.message));

        expect(props.children).toMatchSnapshot();
      });

      test("should prevent memory leaks", async () => {
        const original = async () => {
          await waitFor(50);

          return "done";
        };

        const { unmount } = setup({ original, params: [] });

        unmount();

        await waitFor(75);
      });
    });
  });
});
