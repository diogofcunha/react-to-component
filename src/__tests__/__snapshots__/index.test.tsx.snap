// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToComponent /> Async apis async flow problems should call children with the correct props on failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": null,
        "error": [Error: trolololololololol],
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "trolololololololol",
    },
  ],
}
`;

exports[`<ToComponent /> Async apis async flow problems should call children with the correct props on success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": "I am Diogo and I am 27 years old",
        "error": null,
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loaded",
    },
  ],
}
`;

exports[`<ToComponent /> Async apis async flow problems should call children with the correct props on success and ignore old errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": "I am Diogo and I am 27 years old",
        "error": null,
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loaded",
    },
  ],
}
`;

exports[`<ToComponent /> Async apis should call children with the correct props on failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": null,
        "error": [Error: trolololol],
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "trolololol",
    },
  ],
}
`;

exports[`<ToComponent /> Async apis should call children with the correct props on success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": "I am Diogo and I am 27 years old",
        "error": null,
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loaded",
    },
  ],
}
`;

exports[`<ToComponent /> Sync apis should call children with the correct props on failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": null,
        "error": [Error: Trolololololol],
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Trolololololol",
    },
  ],
}
`;

exports[`<ToComponent /> Sync apis should call children with the correct props on success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": "I am Diogo and I am 27 years old",
        "error": null,
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loaded",
    },
  ],
}
`;

exports[`<ToComponent /> Sync apis should use custom parametersDidChange when provided 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
      },
    ],
    Array [
      Object {
        "data": "I am Diogo and I am 27 years old",
        "error": null,
        "loading": false,
      },
    ],
    Array [
      Object {
        "data": "I am Diogo and I am 27 years old",
        "error": null,
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "Loading",
    },
    Object {
      "isThrow": false,
      "value": "Loaded",
    },
    Object {
      "isThrow": false,
      "value": "Loaded",
    },
  ],
}
`;
